### PR TITLE
[4.0][com_actionlogs] convert to prepared sql

### PR DESCRIPTION
**Pull request by alikon**
### Summary of Changes
use prepared statement for SQL

### Testing Instructions
test com_actionlogs

### Expected result

work as before

### Actual result
N/A

### CODE REVIEW

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on ff662ca4d0155dd463879a1ee16f8422b3b9b3fc<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25671">issues.joomla.org/tracker/joomla-cms/25671</a>.</sub>

File: administrator/components/com_actionlogs/Model/ActionlogsModel.php [hunk 1]
```diff
@@ -147,10 +152,15 @@ protected function getListQuery()
 			// If the chosen range is not more than a year ago
 			if ($date['dNow'] != false)
 			{
+				$dStart = $date['dStart']->format('Y-m-d H:i:s');
+				$dNow   = $date['dNow']->format('Y-m-d H:i:s');
 				$query->where(
-					$db->quoteName('a.log_date') . ' >= ' . $db->quote($date['dStart']->format('Y-m-d H:i:s')) .
-					' AND ' . $db->quoteName('a.log_date') . ' <= ' . $db->quote($date['dNow']->format('Y-m-d H:i:s'))
+					$db->quoteName('a.log_date') . ' >= :dstart' .
+					' AND ' . $db->quoteName('a.log_date') . ' <= :dnow'
```
Comment (Quy) [inline administrator/components/com_actionlogs/Model/ActionlogsModel.php]: Split into 2 `where` statements?

File: administrator/components/com_actionlogs/Model/ActionlogsModel.php [hunk 2]
```diff
@@ -245,13 +260,16 @@ private function buildDateRange($range)
 	 */
 	public function getLogsForItem($extension, $itemId)
 	{
-		$db    = $this->getDbo();
-		$query = $db->getQuery(true)
+		$itemId = (int) $itemId;
+		$db     = $this->getDbo();
+		$query  = $db->getQuery(true)
 			->select('a.*, u.name')
 			->from('#__action_logs AS a')
```
Comment (Quy) [inline administrator/components/com_actionlogs/Model/ActionlogsModel.php]:
```suggestion
			->from($db->quoteName('#__action_logs', 'a')
```

File: administrator/components/com_actionlogs/Model/ActionlogsModel.php [hunk 3]
```diff
@@ -245,13 +260,16 @@ private function buildDateRange($range)
...
 			->innerJoin('#__users AS u ON a.user_id = u.id')
```
Comment (Quy) [inline administrator/components/com_actionlogs/Model/ActionlogsModel.php]: Use join and apply $db->quoteName.

File: administrator/components/com_actionlogs/Model/ActionlogsModel.php (hunk 1 above)
Comment (richard67) [reply administrator/components/com_actionlogs/Model/ActionlogsModel.php]: Maybe even better `$db->quoteName('a.log_date') . ' BETWEEN :dstart AND :dnow'`, could also gain a bit performance theoretically.

File: administrator/components/com_actionlogs/Model/ActionlogsModel.php [hunk 4]
```diff
@@ -245,13 +260,16 @@ private function buildDateRange($range)
 	 */
 	public function getLogsForItem($extension, $itemId)
 	{
-		$db    = $this->getDbo();
-		$query = $db->getQuery(true)
+		$itemId = (int) $itemId;
+		$db     = $this->getDbo();
+		$query  = $db->getQuery(true)
 			->select('a.*, u.name')
```
Comment (Quy) [inline administrator/components/com_actionlogs/Model/ActionlogsModel.php]:
```suggestion
			->select('a.*', $db->quoteName('u.name'))
```

File: administrator/components/com_actionlogs/Model/ActionlogsModel.php [hunk 5]
```diff
@@ -147,10 +152,15 @@ protected function getListQuery()
 			// If the chosen range is not more than a year ago
 			if ($date['dNow'] != false)
 			{
+				$dStart = $date['dStart']->format('Y-m-d H:i:s');
+				$dNow   = $date['dNow']->format('Y-m-d H:i:s');
 				$query->where(
-					$db->quoteName('a.log_date') . ' >= ' . $db->quote($date['dStart']->format('Y-m-d H:i:s')) .
-					' AND ' . $db->quoteName('a.log_date') . ' <= ' . $db->quote($date['dNow']->format('Y-m-d H:i:s'))
+					$db->quoteName('a.log_date') . ' >= :dstart' .
```
Comment (richard67) [inline administrator/components/com_actionlogs/Model/ActionlogsModel.php]: Replace lines 158 and 159 by this one line:
`$db->quoteName('a.log_date') . ' BETWEEN :dstart AND :dnow'`

**Review by richard67 — CHANGES_REQUESTED** [review]

**Comment by richard67** [comment]
I have not tested this item.<br /><br />Will test later again after code review changes.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25671">issues.joomla.org/tracker/joomla-cms/25671</a>.</sub>

File: administrator/components/com_actionlogs/Model/ActionlogsModel.php (hunk 2 above)
Comment (Quy) [inline administrator/components/com_actionlogs/Model/ActionlogsModel.php]:
```suggestion
			->from($db->quoteName('#__action_logs', 'a'))
```

File: administrator/components/com_actionlogs/Model/ActionlogsModel.php (hunk 4 above)
Comment (Quy) [inline administrator/components/com_actionlogs/Model/ActionlogsModel.php]:
```suggestion
			->select('a.*', $db->quoteName('u.name'))
```

File: administrator/components/com_actionlogs/Model/ActionlogsModel.php (hunk 1 above)
Comment (alikon) [reply administrator/components/com_actionlogs/Model/ActionlogsModel.php]: used between

File: administrator/components/com_actionlogs/Model/ActionlogsModel.php (hunk 3 above)
Comment (alikon) [reply administrator/components/com_actionlogs/Model/ActionlogsModel.php]: done

File: administrator/components/com_actionlogs/Model/ActionlogsModel.php (hunk 5 above)
Comment (alikon) [reply administrator/components/com_actionlogs/Model/ActionlogsModel.php]: done

File: administrator/components/com_actionlogs/Model/ActionlogsModel.php [hunk 6]
```diff
@@ -245,13 +259,16 @@ private function buildDateRange($range)
...
-			->from('#__action_logs AS a')
-			->innerJoin('#__users AS u ON a.user_id = u.id')
-			->where($db->quoteName('a.extension') . ' = ' . $db->quote($extension))
-			->where($db->quoteName('a.item_id') . ' = ' . (int) $itemId);
+		$itemId = (int) $itemId;
+		$db     = $this->getDbo();
+		$query  = $db->getQuery(true)
+			->select('a.*', $db->quoteName('u.name'))
+			->from($db->quoteName('#__action_logs', 'a'))
+			->join('INNER', $db->quoteName('#__users', 'u') . ' ON ' . $db->quoteName('a.user_id') . ' = ' . $db->quoteName('u.id))
```
Comment (richard67) [inline administrator/components/com_actionlogs/Model/ActionlogsModel.php]:
```suggestion
			->join('INNER', $db->quoteName('#__users', 'u') . ' ON ' . $db->quoteName('a.user_id') . ' = ' . $db->quoteName('u.id'))
```

File: administrator/components/com_actionlogs/Model/ActionlogsModel.php [hunk 7]
```diff
@@ -318,13 +335,14 @@ private function getLogDataQuery($pks = null)
 	{
 		$db    = $this->getDbo();
 		$query = $db->getQuery(true)
-			->select('a.*, u.name')
-			->from('#__action_logs AS a')
-			->innerJoin('#__users AS u ON a.user_id = u.id');
+			->select('a.*', $db->quoteName('u.name'))
+			->from($db->quoteName('#__action_logs', 'a'))
+			->join('INNER', $db->quoteName('#__users', 'u') . ' ON ' . $db->quoteName('a.user_id') . ' = ' . $db->quoteName('u.id));
```
Comment (richard67) [inline administrator/components/com_actionlogs/Model/ActionlogsModel.php]:
```suggestion
			->join('INNER', $db->quoteName('#__users', 'u') . ' ON ' . $db->quoteName('a.user_id') . ' = ' . $db->quoteName('u.id'));
```

**Review by richard67 — CHANGES_REQUESTED** [review]
Please fix syntax errors.

**Comment by richard67** [comment]
@alikon I get a PHP notice as follows when testing your PR now:

> [Mon Jul 22 19:02:51.665201 2019] [php7:notice] [pid 933] [client 192.168.98.1:50765] PHP Notice:  Undefined property: stdClass::$name in /home/richard/lamp/public_html/joomla-cms-4.0-dev/administrator/components/com_actionlogs/Helper/ActionlogsHelper.php on line 66, referer: https://www.joomla-40-dev.vmnet2.local/administrator/index.php?option=com_actionlogs&view=actionlogs

I will update my 4.0-dev branch to latest changes and set up my testing environment from scratch and test again.

**Comment by alikon** [comment]
Sorry folks for all the issues my pr's raise, I will learn a more user
friendly way to make pr's to lower my shit and reduce the community
efforts, but thanks for your efforts to help


Il lun 22 lug 2019, 19:06 Richard Fath <notifications@github.com> ha
scritto:

**Comment by richard67** [comment]
Just don't code in trains ... especially not in Italian ones :tongue: ... joking ... German trains are even worse ... so I would recommend to use Swiss trains ;-) Ah, and maybe more important: Never code on mobile phone, I never would do that, never. It is simply too small for such works.

**Comment by richard67** [comment]
I have tested this item :red_circle: unsuccessfully on 25c3a3eeae95da7f9cbc3865ccf0acdce5a46f28<br /><br />When doing CSV export of all action log entries, I get a PHP warning as follows:

> PHP Notice:  Undefined property: stdClass::$name in /home/richard/lamp/public_html/joomla-cms-4.0-dev/administrator/components/com_actionlogs/Helper/ActionlogsHelper.php on line 66, referer: https://www.joomla-40-dev.vmnet2.local/administrator/index.php?option=com_actionlogs&view=actionlogs

Without this PR I don't get that.

From reading code I did not find the reason for that yet. Will continue to try to find it, but maybe someone else is faster.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25671">issues.joomla.org/tracker/joomla-cms/25671</a>.</sub>

**Comment by richard67** [comment]
@alikon Beside the PHP notice for which I haven't found the reason yet: Is there any reason why you haven't converted the query here to prepared?
[https://github.com/joomla/joomla-cms/blob/25c3a3eeae95da7f9cbc3865ccf0acdce5a46f28/administrator/components/com_actionlogs/Model/ActionlogsModel.php#L108](https://github.com/joomla/joomla-cms/blob/25c3a3eeae95da7f9cbc3865ccf0acdce5a46f28/administrator/components/com_actionlogs/Model/ActionlogsModel.php#L108)

File: administrator/components/com_actionlogs/Model/ActionlogsModel.php [hunk 8]
```diff
@@ -147,10 +152,14 @@ protected function getListQuery()
...
+				$dStart = $date['dStart']->format('Y-m-d H:i:s');
+				$dNow   = $date['dNow']->format('Y-m-d H:i:s');
 				$query->where(
-					$db->quoteName('a.log_date') . ' >= ' . $db->quote($date['dStart']->format('Y-m-d H:i:s')) .
-					' AND ' . $db->quoteName('a.log_date') . ' <= ' . $db->quote($date['dNow']->format('Y-m-d H:i:s'))
+					$db->quoteName('a.log_date') . ' BETWEEN :dstart AND :dnow'
 				);
+				$query->bind(':dstart', $dStart);
+				$query->bind(':dnow', $dNow);
+
```
Comment (Quy) [inline administrator/components/com_actionlogs/Model/ActionlogsModel.php]:
```suggestion
```

**Comment by richard67** [comment]
@SharkyKZ I tried to test this PR but got a PHP notice, details see my test result above. I've tried to find the reason, but I was not successful. Maybe you can have a look?

File: administrator/components/com_actionlogs/Model/ActionlogsModel.php [hunk 9]
```diff
@@ -414,3 +433,4 @@ public function getFilterForm($data = array(), $loadData = true)
 		return $form;
 	}
 }
+
```
Comment (Quy) [inline administrator/components/com_actionlogs/Model/ActionlogsModel.php]:
```suggestion
```

**Comment by alikon** [comment]
@richard67  there is nothing to be prepared there https://github.com/joomla/joomla-cms/blob/25c3a3eeae95da7f9cbc3865ccf0acdce5a46f28/administrator/components/com_actionlogs/Model/ActionlogsModel.php#L108  (i.e. there are no variables to be binded)

**Comment by richard67** [comment]
Sorry, check line 111 3 lines below the one l’ve linked.

**Comment by alikon** [comment]
if  you mean this `->leftJoin('#__users AS u ON a.user_id = u.id');`  previuos answer is still  valid, maybe you can argue that should be `->join('LEFT`'  and `quoteName` 'd

**Comment by HLeithner** [comment]
I think he mean
```
			->select('a.*, u.name')
			->from('#__action_logs AS a')
			->leftJoin('#__users AS u ON a.user_id = u.id');
		// Get ordering
		$fullorderCol = $this->state->get('list.fullordering', 'a.id DESC');
```
to
```
			->select('a.*,' . $db->quoteName('u.name'))
			->from($db->quoteName('#__action_logs', 'a'))
			->leftJoin($db->quoteName('#__users', 'u'), $db->quoteName('a.user_id') .' = ' . $db->quoteName('u.id'));
		// Get ordering
		$fullorderCol = $this->state->get('list.fullordering', $db->quoteName('a.id') . ' DESC');
```

**Comment by richard67** [comment]
@alikon You are right, nothing to be prepared statement there. My mistake. But as Harald said, quoting would be nice.

**Comment by alikon** [comment]
@richard67  can you confirm that https://github.com/joomla/joomla-cms/pull/25671#issuecomment-513876879  has been fixed by my last commit......  if so i'll update your last cs comment accordindlngly

**Comment by richard67** [comment]
@alikon I will test a bit later after eating. What do you mean with my last cs comment?

**Comment by alikon** [comment]
_buon appetito_
this one https://github.com/joomla/joomla-cms/pull/25671#issuecomment-514146954

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on 82894972e3e6aa7975032e718b4e5bc5d1562534<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25671">issues.joomla.org/tracker/joomla-cms/25671</a>.</sub>

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 82894972e3e6aa7975032e718b4e5bc5d1562534<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25671">issues.joomla.org/tracker/joomla-cms/25671</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25671">issues.joomla.org/tracker/joomla-cms/25671</a>.</sub>

**Comment by alikon** [comment]
pleas  wait 1 sec

File: administrator/components/com_actionlogs/Model/ActionlogsModel.php [hunk 10]
```diff
@@ -105,9 +106,11 @@ protected function getListQuery()
...
 		$db    = $this->getDbo();
 		$query = $db->getQuery(true)
-			->select('a.*, u.name')
-			->from('#__action_logs AS a')
-			->leftJoin('#__users AS u ON a.user_id = u.id');
+			->select('a.*')
+			->select($db->quoteName('u.name'))
+			->from($db->quoteName('#__action_logs', 'a'))
+			->join('LEFT', $db->quoteName('#__users', 'u') . ' ON ' . $db->quoteName('a.user_id') . ' = ' . $db->quoteName('u.id'));
+
```
Comment (Quy) [inline administrator/components/com_actionlogs/Model/ActionlogsModel.php:113]:
```suggestion
```

**Comment by alikon** [comment]
please a quick retest when you can, updated with the last discover

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on 5125c15f6c2c5010536ba8b4628a809f5a49ada6<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25671">issues.joomla.org/tracker/joomla-cms/25671</a>.</sub>

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 4bf53242f30185a77d4086f9d31dc0cee6d4879e<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25671">issues.joomla.org/tracker/joomla-cms/25671</a>.</sub>

**Comment by alikon** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25671">issues.joomla.org/tracker/joomla-cms/25671</a>.</sub>

**Comment by wilsonge** [comment]
Thanks!

**Pull request merged**